### ISSUE TITLE
Quaternion.transform(Vector3) is not thread safe

### ISSUE THREAD

**Issue by theagentd** [libgdx/libgdx#846]
Quaternion.transform(Vector3) is not thread safe since it uses two static variables, tmp1 and tmp2:

```
private static Quaternion tmp1 = new Quaternion(0, 0, 0, 0);
private static Quaternion tmp2 = new Quaternion(0, 0, 0, 0);

public Vector3 transform (Vector3 v) {
    tmp2.set(this);
    tmp2.conjugate();
    tmp2.mulLeft(tmp1.set(v.x, v.y, v.z, 0)).mulLeft(this);

    v.x = tmp2.x;
    v.y = tmp2.y;
    v.z = tmp2.z;
    return v;
}
```

This is very bad practice and I discovered this issue the hard way ("Why are my skeleton animated objects twitching like hell?!"). A fix would be to "hard code" the procedure, which I also believe is slightly faster. The solution I'm using right now can be found here:

http://pastie.org/8414910

It may need some very light reworking to fit the style of LibGDX.

**Comment by JBraig** [comment]
There doesn't seem to have been any progress regarding this issue in the last seven years, but I think the topic deserves another look after all this time. I recently stumbled over this issue by accident, and while I know that libgdx isn't thread safe unless explicitly specified, this is a level above what I'd usually think of when thinking of thread safety - I only discovered this problem when independent Matrix4-instances in different threads started interfering with each other. Some functions like Matrix4.mulLeft() should be fairly straightforward to unroll, for others I'm not so sure. According to the [old, now archived forum post,](https://web.archive.org/web/20150920150715/http://www.badlogicgames.com/forum/viewtopic.php?f=23&t=7814) this was done to relieve GC pressure, and there's some mention of modifying the API so users can pass their own vectors, matrices or whatever instead of using the temp ones, yet this seems to never have been implemented. Any specific reason for that?
